### PR TITLE
8286125: C2: "bad AD file" with PopulateIndex on x86_64

### PR DESCRIPTION
**Pull request by pfustc**
A fuzzer test reports an assertion failure issue with PopulateIndexNode
on x86_64. It can be reproduced by the new jtreg case inside this patch.
Root cause is that C2 superword creates a PopulateIndexNode by mistake
while vectorizing below loop.
```
    for (int i = 304; i > 15; i -= 3) {
        int c = 16;
        do {
            for (int t = 1; t < 1; t++) {}
            arr[c + 1] >>= i;
        } while (--c > 0);
    }
```
This is a corner loop case with redundant code inside. After several C2
optimizations, the do-while loop inside is unrolled and then isomorphic
right shift statements can be combined in the superword optimization.
Since all shift counts are the same loop IV value `i`, superword should
generate a RShiftCntVNode to create a vector of scalar replications of
the loop IV. But after JDK-8280510, a PopulateIndexNode is generated by
mistake because of the `opd == iv()` condition.

To fix this, we add a `have_same_inputs` condition here checking if all
inputs at position `opd_idx` of nodes in the pack are the same. If true,
C2 code should NOT run into this block to generate a PopulateIndexNode.
Instead, it should run into the next block for scalar replications.

Additionally, only adding this condition here is still not good enough
because it breaks the experimental post loop vectorization. As in post
loops, all packs are singleton, i.e., `have_same_inputs` is always true.
Hence, we also add a pack size check here to make post loop logic run
into this block. It's safe to let it go because post loop never needs
scalar replications of the loop IV - it never combines nodes in packs.

We also add two more assertions in the code.

Jtreg hotspot::hotspot_all_no_apps, jdk::tier1~3 and langtools::tier1
are tested and no issue is found.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8286125](https://bugs.openjdk.java.net/browse/JDK-8286125): C2: "bad AD file" with PopulateIndex on x86_64

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [e91eb0bf](https://git.openjdk.java.net/jdk/pull/8587/files/e91eb0bf851787d81958b20b4b1bb6fedda17fc4)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [3af03a80](https://git.openjdk.java.net/jdk/pull/8587/files/3af03a806ba5d37425785907788a11a35a61f611)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8587/head:pull/8587` \
`$ git checkout pull/8587`

Update a local copy of the PR: \
`$ git checkout pull/8587` \
`$ git pull https://git.openjdk.java.net/jdk pull/8587/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8587`

View PR using the GUI difftool: \
`$ git pr show -t 8587`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8587.diff">https://git.openjdk.java.net/jdk/pull/8587.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Seems reasonable. Let me test it.

**Comment by vnkozlov** [comment]
Meanwhile, please, update it to latest JDK.

**Review by vnkozlov — APPROVED** [review]
My tier1-4 testing passed clean.

**Comment by pfustc** [comment]
> Meanwhile, please, update it to latest JDK.

Thanks @vnkozlov for looking at this. Patch is rebased to JDK master.

**Comment by vnkozlov** [comment]
You need second review. It is not trivial.

File: test/hotspot/jtreg/compiler/vectorization/TestReplicateLoopIV.java
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/vectorization/TestReplicateLoopIV.java:1]:
```suggestion

```

**Review by TobiHartmann — COMMENTED** [review]
> As in post loops, all packs are singleton, i.e., have_same_inputs is always true.
Hence, we also add a pack size check here to make post loop logic run
into this block. It's safe to let it go because post loop never needs
scalar replications of the loop IV - it never combines nodes in packs.

I don't understand why we need to execute that code for post loops. Could you elaborate why the `p->size() == 1` check is needed?

**Comment by pfustc** [comment]
Hi @TobiHartmann ,

Thanks for your review.

> I don't understand why we need to execute that code for post loops. Could you elaborate why the `p->size() == 1` check is needed?

In post loop vectorization, superword doesn't combine multiple scalar nodes to one vector node. Instead, each scalar node is transformed to a vector node with vector mask. If a loop body statement has the induction variable involved, such as
```
for (int i = 0; i < length; i++) {
    a[i] = i;
}
```
then superword should run into this `if (opd == iv() && ...) {...}` block to create an index vector for vectorizing post loop. But packs in post loops are special (singletons) - each has only one scalar node inside. So `same_inputs(...)` always returns true. If we just add the `have_same_inputs` condition, post loop vectorization will generate incorrect result because index vectors cannot be created. Hence we add another `p->size() == 1` check here.

**Comment by TobiHartmann** [comment]
Thanks for the clarification. Makes sense.

**Review by TobiHartmann — APPROVED** [review]
Please remove the newline from the beginning of the test before integrating.

**Pull request closed without merging**
